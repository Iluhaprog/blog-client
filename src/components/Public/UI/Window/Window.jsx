import React, {useEffect, useState} from 'react';
import styled from 'styled-components';
import {Titlebar} from '../Titlebar';
import {WindowFrame} from './Frame';
import PropTypes from 'prop-types';

const WindowBox = styled.div`
  position: absolute;
  width: fit-content;
`;

const Window = ({title, onClose, children}) => {
  const [x, setX] = useState(100);
  const [y, setY] = useState(100);
  const [offset, setOffset] = useState({x: 0, y: 0});
  const [isMouseDown, setIsMouseDown] = useState(false);

  const mouseDown = (e) => {
    setOffset({
      x: e.target.offsetLeft - e.clientX,
      y: e.target.offsetTop - e.clientY,
    });
    setIsMouseDown(true);
  };
  const mouseUp = () => {
    setIsMouseDown(false);
  };

  const move = (e) => {
    setX(e.clientX + offset.x);
    setY(e.clientY + offset.y);
  };

  useEffect(() => {
    if (isMouseDown) {
      document.onmousemove = move;
      document.onmouseup = mouseUp;
    } else {
      document.onmousemove = null;
    }
  }, [isMouseDown]);

  return (
    <WindowBox
      style={{
        top: `${y}px`,
        left: `${x}px`,
      }}
    >
      <Titlebar
        onMouseDown={mouseDown}
        onMouseUp={mouseUp}
        isActive={true}
        title={title}
        onClose={onClose}
      />
      <WindowFrame>
        {children}
      </WindowFrame>
    </WindowBox>
  );
};

Window.propTypes = {
  title: PropTypes.string,
  onClose: PropTypes.func,
  children: PropTypes.node,
};

export {Window};
