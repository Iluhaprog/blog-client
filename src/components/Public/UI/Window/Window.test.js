import {render} from '@testing-library/react';
import {Window} from './Window';

test('Should render window', () => {
  const {asFragment} = render(
    <Window title={'TEST'}>
      <p>TEST BODY</p>
    </Window>
  );
  expect(asFragment()).toMatchSnapshot();
});
