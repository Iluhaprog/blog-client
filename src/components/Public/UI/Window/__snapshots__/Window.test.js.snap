// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render window 1`] = `
<DocumentFragment>
  <div
    class="sc-jSgupP iSPhCi"
    style="top: 100px; left: 100px;"
  >
    <div
      class="sc-dlfnbm gSaEMk"
    >
      <div
        class="sc-hKgILt dfsasS"
      >
        <button
          class="sc-bdfBwQ hUrxdp"
        >
          <div
            class="sc-gsTCUz kbJdGI"
            height="20"
            width="20"
          >
            <svg>
              close.svg
            </svg>
          </div>
        </button>
      </div>
      <p>
        TEST
      </p>
      <div
        class="sc-hKgILt dfsasS"
      />
    </div>
    <div
      class="sc-eCssSg hnaOHp"
    >
      <p>
        TEST BODY
      </p>
    </div>
  </div>
</DocumentFragment>
`;
